### PR TITLE
PLAT-1260-4 Implement user specific preferred popup placement

### PR DESCRIPTION
**Pull request by alexander-schmidt**
- Popups can now be placed/moved relative to their respective `IDomPopupContext`.
- Implemented used-specific selection of preferred popup placements.
  - Implemented three new context-relative strategies.
  - Those and the good old "show popup at click coordinates" strategy (which is still the default) are now available for selection in the user profile.
- Renamed everything "popup position" to "popup placement".

Peek 2023-01-11 15-29

### CODE REVIEW

File: platform-ajax/src/main/typescript/com/softicar/platform/ajax/engine/Popup.ts [hunk 1]
```diff
@@ -78,6 +44,72 @@ class DomPopupEngine {
...
+			case "CENTER": popupTop = popupHeight/2; break;
+			case "BOTTOM": popupTop = popupHeight; break;
+		}
+
+		// move popup
+		popupFrame.style.left = `calc(${x}% - ${popupLeft}px)`;
+		popupFrame.style.top = `calc(${y}% - ${popupTop}px)`;
+	}
+
+	private movePopupByPixels(parent: HTMLElement, popupFrame: HTMLElement, x: number, y: number, xAlign: string, yAlign: string) {
```
Comment (alexander-schmidt) [inline platform-ajax/src/main/typescript/com/softicar/platform/ajax/engine/Popup.ts:72]: This was bluntly extracted.

**Review by oliver-richers — DISMISSED** [review]

File: platform-ajax/src/main/typescript/com/softicar/platform/ajax/engine/Popup.ts [hunk 2]
```diff
@@ -78,6 +44,72 @@ class DomPopupEngine {
 	private raise(popupFrame: HTMLElement) {
 		AJAX_ENGINE.raise(popupFrame);
 	}
+
+	private movePopupByPercent(popupFrame: HTMLElement, x: number, y: number, xAlign: string, yAlign: string) {
+		// compute left
+		let popupWidth = popupFrame.offsetWidth;
+		let popupLeft = 0;
```
[the remark this answers is not in the document]
Comment (alexander-schmidt) [reply platform-ajax/src/main/typescript/com/softicar/platform/ajax/engine/Popup.ts:51]: okay

File: platform-core-module/src/main/java/com/softicar/platform/core/module/user/UserPreferencesPreferredPopupPlacement.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+		return displayName;
+	}
+
+	/**
+	 * Returns the associated {@link IDomPopupPlacementStrategy}.
+	 *
+	 * @return the {@link IDomPopupPlacementStrategy} (never <i>null</i>)
+	 */
+	public IDomPopupPlacementStrategy get() {
```
[the remark this answers is not in the document]
Comment (alexander-schmidt) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/user/UserPreferencesPreferredPopupPlacement.java:52]: okay

File: platform-dom/src/main/java/com/softicar/platform/dom/DomI18n.java [hunk 4]
```diff
@@ -28,8 +28,14 @@ public interface DomI18n extends CommonCoreI18n {
 		.de("Sollen dieses Fenster und alle Unterfenster wirklich geschlossen werden?");
 	I18n0 ARE_YOU_SURE_TO_CLOSE_THIS_WINDOW_QUESTION = new I18n0("Are you sure to close this window?")//
 		.de("Soll dieses Fenster wirklich geschlossen werden?");
+	I18n0 AT_CLICK_POSITION = new I18n0("At click position")//
+		.de("Bei Klick-Position");
+	I18n0 BELOW_CLICKED_ELEMENT = new I18n0("Below clicked element")//
```
[the remark this answers is not in the document]
Comment (alexander-schmidt) [reply platform-dom/src/main/java/com/softicar/platform/dom/DomI18n.java]: already removed

File: platform-dom/src/main/java/com/softicar/platform/dom/DomI18n.java [hunk 5]
```diff
@@ -28,8 +28,14 @@ public interface DomI18n extends CommonCoreI18n {
 		.de("Sollen dieses Fenster und alle Unterfenster wirklich geschlossen werden?");
 	I18n0 ARE_YOU_SURE_TO_CLOSE_THIS_WINDOW_QUESTION = new I18n0("Are you sure to close this window?")//
 		.de("Soll dieses Fenster wirklich geschlossen werden?");
+	I18n0 AT_CLICK_POSITION = new I18n0("At click position")//
```
[the remark this answers is not in the document]
Comment (alexander-schmidt) [reply platform-dom/src/main/java/com/softicar/platform/dom/DomI18n.java:31]: might be confusing for touch interfaces; leaving as-is

File: platform-dom/src/main/java/com/softicar/platform/dom/elements/popup/placement/strategy/AbstractDomPopupViewportAlignmentPlacementStrategy.java [hunk 6]
```diff
@@ -1,20 +1,21 @@
...
+import com.softicar.platform.dom.elements.popup.placement.DomPopupPlacement;
+import com.softicar.platform.dom.engine.DomPopupOffsetUnit;
 import com.softicar.platform.dom.engine.DomPopupXAlign;
 import com.softicar.platform.dom.engine.DomPopupYAlign;
 import com.softicar.platform.dom.event.IDomEvent;
 import com.softicar.platform.dom.style.CssPercent;
 import java.util.Optional;
 
-abstract class AbstractDomPopupViewportAlignmentPositionStrategy implements IDomPopupPositionStrategy {
+abstract class AbstractDomPopupViewportAlignmentPlacementStrategy implements IDomPopupPlacementStrategy {
```
[the remark this answers is not in the document]
Comment (alexander-schmidt) [reply platform-dom/src/main/java/com/softicar/platform/dom/elements/popup/placement/strategy/AbstractDomPopupViewportAlignmentPlacementStrategy.java:11]: okay

File: platform-dom/src/main/java/com/softicar/platform/dom/elements/popup/placement/strategy/DomPopupBelowEventNodePlacementStrategy.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+	@Override
+	public DomPopupPlacement getPlacement(IDomEvent event) {
+
+		if (event != null) {
+			var rect = event.getBoundingClientRect();
+			int x = Double.valueOf(rect.getX() + rect.getWidth()).intValue();
+			int y = Double.valueOf(rect.getY() + rect.getHeight()).intValue();
+			return new DomPopupPlacement(x, y, DomPopupOffsetUnit.PIXELS, DomPopupXAlign.RIGHT, DomPopupYAlign.TOP);
+		} else {
+			return new DomPopupViewportCenterPlacementStrategy(CssPercent._50, CssPercent._25).getPlacement(event);
```
[the remark this answers is not in the document]
Comment (alexander-schmidt) [reply platform-dom/src/main/java/com/softicar/platform/dom/elements/popup/placement/strategy/DomPopupBelowEventNodePlacementStrategy.java:28]: _null_ is allowed, as per `IDomPopupPlacementStrategy.getPlacement` javadoc. Now passing _null_ explicitly there.

File: platform-dom/src/main/java/com/softicar/platform/dom/elements/popup/placement/strategy/DomPopupEventCoordinatesPlacementStrategy.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+		if (event != null) {
+			return new DomPopupPlacement(//
+				(int) Math.round(event.getClientX()),
+				(int) Math.round(event.getClientY()),
+				DomPopupOffsetUnit.PIXELS,
+				DomPopupXAlign.LEFT,
+				DomPopupYAlign.TOP);
+		} else {
+			return new DomPopupViewportCenterPlacementStrategy(CssPercent._50, CssPercent._25).getPlacement(event);
```
[the remark this answers is not in the document]
Comment (alexander-schmidt) [reply platform-dom/src/main/java/com/softicar/platform/dom/elements/popup/placement/strategy/DomPopupEventCoordinatesPlacementStrategy.java:29]: _null_ is allowed, as per `IDomPopupPlacementStrategy.getPlacement` javadoc. Now passing _null_ explicitly there.

**Review by oliver-richers — APPROVED** [review]

**Pull request merged**